src/renderer/d3d.rs

```diff
@@ -10,186 +7,122 @@ use skia_safe::{
-        Err(())
-    }
-}
+fn get_hardware_adapter(factory: &IDXGIFactory4) -> Result<IDXGIAdapter1> {
+    for i in 0.. {
+        let adapter = unsafe { factory.EnumAdapters1(i)? };
+        let mut desc = Default::default();
+        unsafe {
+            adapter.GetDesc1(&mut desc)?;
+        }
 
-fn is_adapter_suitable(adapter: *mut IDXGIAdapter1) -> bool {
-    let mut desc = DXGI_ADAPTER_DESC1::default();
-    if SUCCEEDED(unsafe { (*adapter).GetDesc1(&mut desc) }) {
-        if (desc.Flags & DXGI_ADAPTER_FLAG_SOFTWARE) == DXGI_ADAPTER_FLAG_SOFTWARE {
-            // Don't select the Basic Render Driver adapter.
-            false
-        } else {
-            // Check to see whether the adapter supports Direct3D 12, but don't create the
-            // actual device yet.
-            unsafe {
-                SUCCEEDED(D3D12CreateDevice(
-                    adapter as *mut IUnknown,
-                    D3D_FEATUREL_LEVEL,
-                    &ID3D12Device::uuidof(),
-                    null_mut(),
-                ))
-            }
+        if (DXGI_ADAPTER_FLAG(desc.Flags as i32) & DXGI_ADAPTER_FLAG_SOFTWARE)
+            != DXGI_ADAPTER_FLAG_NONE
+        {
+            continue;
         }
-    } else {
-        false
-    }
-}
 
-fn find_first_suitable(
-    enumerator: &dyn Fn(u32) -> Result<ComPtr<IDXGIAdapter1>, ()>,
-) -> Result<ComPtr<IDXGIAdapter1>, ()> {
-    let mut adapter_index = 0;
-    loop {
-        if let Ok(adapter) = enumerator(adapter_index) {
-            if is_adapter_suitable(adapter.as_raw()) {
-                break Ok(adapter);
+        unsafe {
+            if D3D12CreateDevice(
+                &adapter,
+                D3D_FEATURE_LEVEL_11_0,
+                &mut Option::<ID3D12Device>::None,
+            )
+            .is_ok()
+            {
+                return Ok(adapter);
             }
-        } else {
-            break Err(());
         }
-        adapter_index += 1;
     }
-}
 
-// Helper function for acquiring the first available hardware adapter that supports Direct3D 12.
-// If no such adapter can be found, *ppAdapter will be set to nullptr.
-fn get_hardware_adapter(factory: &ComPtr<IDXGIFactory4>) -> Result<ComPtr<IDXGIAdapter1>, ()> {
-    let adapter = if let Ok(factory6) = factory.cast::<IDXGIFactory6>() {
-        find_first_suitable(&|index: u32| -> Result<ComPtr<IDXGIAdapter1>, ()> {
-            call_com_fn(|adapter, id| unsafe {
-                factory6.EnumAdapterByGpuPreference(
-                    index,
-                    DXGI_GPU_PREFERENCE_HIGH_PERFORMANCE,
-                    id,
-                    adapter,
-                )
-            })
-        })
-    } else {
-        Err(())
-    };
-
-    if adapter.is_err() {
-        find_first_suitable(&|index: u32| -> Result<ComPtr<IDXGIAdapter1>, ()> {
-            call_com_fn(|adapter, _| unsafe { factory.EnumAdapters(index, adapter) })
-        })
-    } else {
-        adapter
-    }
+    unreachable!()
```

Comment:
I think you can put a comment here that the function returns when `EnumAdapters` has enumerated all the adapters and return a failure, so this code is never reached.